mambaSharedFramework/Pantos-Generic Playlist Parsing/Pantos-Generic Tag Validators/EXT_X_DATERANGEPlaylistValidator.swift

```diff
@@ -0,0 +1,98 @@
+//
+//  EXT_X_DATERANGEPlaylistValidator.swift
+//  mamba
+//
+//  Created by Robert Galluccio on 28/03/2020.
+//  Copyright © 2020 Comcast Corporation.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License. All rights reserved.
+//
+
+import Foundation
+
+class EXT_X_DATERANGEPlaylistValidator: VariantPlaylistValidator {
+    
+    static func validate(variantPlaylist: VariantPlaylistInterface) -> [PlaylistValidationIssue] {
+        
+        var programDateTimeTagsCount = 0
+        var daterangeTags = [PlaylistTag]()
+        for tag in variantPlaylist.tags {
+            if tag.tagDescriptor == PantosTag.EXT_X_PROGRAM_DATE_TIME {
+                programDateTimeTagsCount += 1
+            } else if tag.tagDescriptor == PantosTag.EXT_X_DATERANGE {
+                daterangeTags.append(tag)
+            }
+        }
+        
+        var validationIssues = [PlaylistValidationIssue]()
+        validationIssues.append(contentsOf: validateProgramDateTime(programDateTimeTagsCount: programDateTimeTagsCount, daterangeTagsCount: daterangeTags.count))
+        validationIssues.append(contentsOf: validateMultipleTags(daterangeTags: daterangeTags))
+        
+        return validationIssues
+    }
+    
+    // If a Playlist contains an EXT-X-DATERANGE tag, it MUST also contain
+    // at least one EXT-X-PROGRAM-DATE-TIME tag.
+    private static func validateProgramDateTime(programDateTimeTagsCount: Int, daterangeTagsCount: Int) -> [PlaylistValidationIssue] {
+        
+        guard daterangeTagsCount > 0, programDateTimeTagsCount == 0 else {
+            return []
+        }
+        return [PlaylistValidationIssue(description: .EXT_X_DATERANGEExistsWithNoEXT_X_PROGRAM_DATE_TIME, severity: .error)]
```

Comment:
As stated in a comment, these should probably be `warning` since they are not fatal to playback.